File: CS50x_2022-2023/Week_6/sentimental-readability/readability.py
from cs50 import get_string
from sys import exit
from decimal import Decimal


def main():
    # Get string, count letters, and words, and sentences
    str = get_string("Text: ")
    letters = count_letters(str)
    words = count_words(str)
    sentences = count_sentences(str)

    # Calculate letters-per-words and sentences-per-words
    letters_per_words = (letters / words) * 100
    sentences_per_words = (sentences / words) * 100

    # Apply Coleman-Liau index
    index = Decimal(0.0588 * letters_per_words - 0.296 * sentences_per_words - 15.8)
    index = round(index)

    # print(f"{letters} {words} {sentences}")

    # Special conditions for >16 and <1 index
    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")

    exit(0)


def count_letters(text):

    count = 0
    # for (int i = 0, n = strlen(text); i < n; i++)
    for i in text:
        if i.isalpha():
            count += 1
    # Return count of letters
    return count


def count_words(text):

    count = 0
    n = len(text)
    # Same for function but in more C way
    for i in range(n):
        if text[i] == " " or i == n - 1:
            count += 1
    # Return count of words
    return count


def count_sentences(text):

    count = 0

    for i in text:
        if i == "." or i == "!" or i == "?":
            count += 1

    return count


main()

File: CS50x_2022-2023/Week_2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

int count_letters(string text);
//A function to find number of letters in given string.
int count_words(string text);
//A function to find number of words in given string.
int count_sentences(string text);
//A function to find number of sentences in given string.

int main(void)
{
    string n = get_string("Text: ");
    int letters, words, sentences;
    float index, letters_per_words, sentences_per_words;
    //Because the Coleman-Liau index can have decimal results, we store those variables as floats.

    //printf("%s\n", n);

    letters = count_letters(n);
    words = count_words(n);
    sentences = count_sentences(n);

    letters_per_words = ((float) letters / words) * 100;
    sentences_per_words = ((float) sentences / words) * 100;

    index = round(0.0588 * letters_per_words - 0.296 * sentences_per_words - 15.8);
    //The Coleman-Liau index

    //printf("%i letters\n", letters);
    //printf("%i words\n", words);
    //printf("%i sentences\n", sentences);

    if (index > 16)
    {
        printf("Grade 16+\n");
    }//If the result of index is bigger than 16, "16+" will be printed.
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %.0f\n", index);
    }//Else, the rounded result will be printed.
}

int count_letters(string text)
{
    int letter = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (islower(text[i]) || isupper(text[i]))
        {
            letter++;
        }//If the current character of the string is a lowercase or a uppercase character, letter counter will increase.
    }

    return letter;
}

int count_words(string text)
{
    int word = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isspace(text[i]) || i == n - 1)
        {
            word++;
        }
    }//If the current character of the string is a whitespace character, or is the latest character, word counter will increase.

    return word;
}

int count_sentences(string text)
{
    int sentence = 0;

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentence++;
        }
    }//If the current character of the string is a dot, exclamation point, or question mark, sentence counter will increase.

    return sentence;
}